### ISSUE TITLE
[NEWS] Obsługa zamówień z odroczoną płatnością dla firm / Processing of orders with Extended Payment Terms

### ISSUE THREAD

**Issue by MaciejFrackowiak** [allegro/allegro-api#4100]
Już w **lutym 2021**, wraz ze startem [Allegro Biznes](https://allegro.pl/dla-sprzedajacych/zmiany-na-allegro-od-8-lutego-k1xWm7dX1ue#allegro-biznes-najlepsze-miejsce-na-firmowe-zakupy-i-sprzedaz), udostępnimy nową metodę płatności dla transakcji pomiędzy firmami (B2B) - **Odroczona płatność dla Firm**. Sprzedający, który chce udostępnić odroczoną płatność w swoich ofertach, musi mieć konto firmowe oraz spełnić warunki opisane w [Pomocy Allegro](https://allegro.pl/pomoc/dla-sprzedajacych/platnosc-od-kupujacego/odroczona-platnosc-dla-firm-informacje-dla-sprzedajacych-k1WOEaoXZHL#kiedy-udostepnisz-odroczona-platnosc-w-swoich-ofertach). 

W związku z wprowadzeniem odroczonej płatności dla zamówienia w statusie **“READY_FOR_PROCESSING”** w:
- [GET /order/checkout-forms/{id}](https://developer.allegro.pl/documentation/#operation/getOrdersDetailsUsingGET) w obiekcie **payment**:
  - udostępnimy nowy typ płatności - **“type”: “EXTENDED_TERM”**,
  - udostępnimy nowego operatora płatności - **“provider”: “EPT”**,
  - pola **“finishedAt”** oraz **“paidAmount”** przyjmą wartość **null** do momentu faktycznego przekazania środków przez dostawcę usługi (aby przelew został zrealizowany, sprzedający [musi wystawić fakturę VAT oraz dodać numer przesyłki do zamówienia](https://allegro.pl/pomoc/dla-sprzedajacych/platnosc-od-kupujacego/odroczona-platnosc-dla-firm-informacje-dla-sprzedajacych-k1WOEaoXZHL#kiedy-i-jak-otrzymasz-pieniadze-za-sprzedany-przedmiot)). 

  Przykładowy widok obiektu **payment** dla zamówień z odroczoną płatnością przed  otrzymaniem płatności:
  ```
  "payment": {
       "id": "b83c6ea0-3ed0-11eb-a57e-ebbabb38ee2a",
       "type": "EXTENDED_TERM",
       "provider": "EPT",
       "finishedAt": null,
       "paidAmount": null
   }
  ```

  Gdy wpłata trafi na konto sprzedającego pola **“finishedAt”** oraz **“paidAmount”** uzupełnimy wartościami:
  ```
  "payment": {
       "id": "b83c6ea0-3ed0-11eb-a57e-ebbabb38ee2a",
       "type": "EXTENDED_TERM",
       "provider": "EPT",
       "finishedAt": "2021-01-18T13:18:04.546Z",
       "paidAmount": {
           "amount": "1150.00",
           "currency": "PLN"
       }
   }
  ```

  **Ważne!** Jeżeli kupujący wybierze odroczoną płatność jako metodę płatności, to tym samym nie będzie mógł już jej anulować i ponownie wypełnić formularza pozakupowego.

- [GET /order/events](https://developer.allegro.pl/documentation/#operation/getOrderEventsUsingGET) zwrócimy dla tego zamówienia **drugie zdarzenie “READY_FOR_PROCESSING”** w momencie przekazania środków przez dostawcę - pole **“occurredAt”** otrzyma jako wartość datę tej operacji (dzięki temu rozpoznasz, że nie jest to duplikat zdarzenia).

Więcej informacji o odroczonej płatności dla firm znajdziesz w [Pomocy Allegro](https://allegro.pl/pomoc/dla-sprzedajacych/platnosc-od-kupujacego/odroczona-platnosc-dla-firm-informacje-dla-sprzedajacych-k1WOEaoXZHL).

-----

Already in **February 2021**, with the launch of [Allegro Biznes](https://allegro.pl/dla-sprzedajacych/zmiany-na-allegro-od-8-lutego-k1xWm7dX1ue#allegro-biznes-najlepsze-miejsce-na-firmowe-zakupy-i-sprzedaz), we will provide a new payment method for transactions between companies (B2B) - **Extended Payment Terms**. A seller who wants to provide Extended Payment Terms in their offers must have a business account and meet the conditions described in [Allegro Help](https://allegro.pl/pomoc/dla-sprzedajacych/platnosc-od-kupujacego/odroczona-platnosc-dla-firm-informacje-dla-sprzedajacych-k1WOEaoXZHL#kiedy-udostepnisz-odroczona-platnosc-w-swoich-ofertach). 

Due to the introduction of Extended Payment Terms for the order in the **“READY_FOR_PROCESSING”** status in:
- [GET /order/checkout-forms/{id}](https://developer.allegro.pl/en/documentation/#operation/getOrdersDetailsUsingGET) in the **payment** object:
  - we will provide a new payment type - **“type”: “EXTENDED_TERM”**,
  - we will provide a new payment provider - **“provider”: “EPT”**,
  - fields **“finishedAt”** and **“paidAmount”** will take a value **null** until the funds are actually transferred by the service provider (for the transfer to be processed, the seller [must issue a VAT invoice and add a tracking number to the order](https://allegro.pl/pomoc/dla-sprzedajacych/platnosc-od-kupujacego/odroczona-platnosc-dla-firm-informacje-dla-sprzedajacych-k1WOEaoXZHL#kiedy-i-jak-otrzymasz-pieniadze-za-sprzedany-przedmiot)). 

  Sample view of **payment** object for orders with Extended Payment Terms before receiving the payment:
  ```
  "payment": {
       "id": "b83c6ea0-3ed0-11eb-a57e-ebbabb38ee2a",
       "type": "EXTENDED_TERM",
       "provider": "EPT",
       "finishedAt": null,
       "paidAmount": null
   }
  ```

  Once the payment will be transferred to the seller`s account, fields **“finishedAt”** and **“paidAmount”** will be updated with values:
  ```
  "payment": {
       "id": "b83c6ea0-3ed0-11eb-a57e-ebbabb38ee2a",
       "type": "EXTENDED_TERM",
       "provider": "EPT",
       "finishedAt": "2021-01-18T13:18:04.546Z",
       "paidAmount": {
           "amount": "1150.00",
           "currency": "PLN"
       }
   }
  ```

  **Note!** If the buyer chooses Extended Payment Terms as the payment method, he will no longer be able to cancel it and fill out the shipping form again.

- [GET /order/events](https://developer.allegro.pl/en/documentation/#operation/getOrderEventsUsingGET) we will return for this order **second event “READY_FOR_PROCESSING”** when the funds are transferred by the provider - field **“occurredAt”** will get the date of this operation as value (this will allow you to recognize that it is not a duplicate event).

You can find more information about Extended Payment Terms in [Allegro Help](https://allegro.pl/pomoc/dla-sprzedajacych/platnosc-od-kupujacego/odroczona-platnosc-dla-firm-informacje-dla-sprzedajacych-k1WOEaoXZHL).

**Comment by SebastianOzdoba** [comment]
@MaciejFrackowiak czy będzie jakiś dedykowany endpoint dla tych faktur by je przekazywać automatycznie? jaki format będzie tej faktury, zakładam że pdf-a bedzie można dostarczyć

**Comment by MaciejFrackowiak** [comment]
@SebastianOzdoba Tak, udostępnimy zasób pozwalający na przekazywanie faktur, będą one w formacie *.pdf.

**Comment by Zbigniew-L** [comment]
Witam, czy znany jest może przypuszczalny termin uruchomienia zasobu do uploadu faktur? Niektórzy nasi klienci się martwią, że bez takiego automatu będą tracić dużo czasu na ręczny upload...

**Comment by MaciejFrackowiak** [comment]
@Zbigniew-L Jak tylko wystartuje Allegro Biznes, zasoby będą dostępne publicznie - [ich kształt jest już znany](https://github.com/allegro/allegro-api/issues/4130).

**Comment by MACskow** [comment]
Witam.
Czy jest jakiś parametr w którym w zamówieniu przekazywana jest ilość dni wybrana przez Klienta ?

**Comment by MaciejFrackowiak** [comment]
@MACskow Nie, obecnie nie ma takiego parametru. Sugestia dot. jego wprowadzenia znajduje się już na [Discussions](https://github.com/allegro/allegro-api/discussions/4333).
